### PR TITLE
[Calling] Reaction mobile view button animations

### PR DESCRIPTION
**Pull request by Mohtasim**
# What

* The changes include a mobile drawer button which has 5 more sub buttons for reaction send
* The component also sends reaction upon click of those button

# Why

# How Tested

* 
Screenshot 2024-03-11 at 4 33 54 PM

* Figma: https://www.figma.com/file/72qJ03EOQExf12g5JTdWjQ/Web-UI-Library?type=design&node-id=11082-10105&mode=design&t=imu9PJpIhmxpTqjv-0
# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/react-components/package.json [hunk 1]
```diff
@@ -62,6 +62,7 @@
     "uuid": "^9.0.0"
   },
   "peerDependencies": {
+    "@azure/communication-calling": "1.23.1-beta.2 || ^1.22.1",
```
Comment (JamesBurnside) [inline packages/react-components/package.json:65]: This package is intentionally agnostic of ACS, instead recreate types. Some type interfaces will look the same, others will need to be reshaped to be an agnostic UI prop.
Looks like this is just for "Reaction" which can recreate as appropriate

File: packages/react-components/src/components/Drawer/DrawerMenu.tsx [hunk 2]
```diff
@@ -104,28 +104,51 @@ export const _DrawerMenu = (props: _DrawerMenuProps): JSX.Element => {
...
-          <DrawerMenuItem
-            {...item}
-            key={`${item.itemKey}` + `${i + 1}`}
-            onItemClick={(ev, itemKey) => {
-              onItemClick(item, ev, itemKey);
-            }}
-          />
-        ))}
-      </Stack>
+      {menuItemsToRender !== undefined && menuItemsToRender[0].itemKey === 'reactions' && (
```
Comment (JamesBurnside) [inline packages/react-components/src/components/Drawer/DrawerMenu.tsx:107]: Using itemKey to pivot the UI behavior is a little hacky, there is nothing in the interface of `<DrawerMenu />` to suggest that using a specific itemKey will translate to new UI.

For interest of time you can keep this this way, but consider an alternative that would scale to future features that may also want custom behavior as a menu item.

One way would be to expose an onRender function. So `_DrawerMenuItemProps` would have an additional optional property called `onRenderMenuItem` that functions as an override. You can scan the codebase for other onRender functions to get an example of how it works, but conceptually it takes in a react component as the prop and renders that instead. The parent of the DrawerMenu would then pass in a react component that would be rendered instead of a default menu item.

Doing it this way then also keeps the drawer menu interface clean and completely agnostic of the specialized way we use it. (Think, if we ever released DrawerMenu as a pure UI component, most scenarios would not involve Reactions. This is the same reason it doesn't take anything like onHoldClick even though sometimes it has a menu item that performs a hold function)

File: packages/react-components/src/components/Drawer/DrawerMenu.tsx [hunk 3]
```diff
@@ -104,28 +104,51 @@ export const _DrawerMenu = (props: _DrawerMenuProps): JSX.Element => {
...
+              {...item}
+              key={`${item.itemKey}` + '0'}
+              shouldFocusOnMount={true}
+              styles={modifiedFirstItemStyle}
+              onItemClick={(ev, itemKey) => {
+                onItemClick(item, ev, itemKey);
+              }}
+            />
+          ))}
+          {menuItemsToRender?.slice(1).map((item, i) => (
```
Comment (JamesBurnside) [inline packages/react-components/src/components/Drawer/DrawerMenu.tsx:141]: This chunk is indentical on each side of the if, only wrap what actually changes in the if-else (or ternary operator in this case).

I _think_ the only difference is not passing in an onItemClick? in which case just make the itemClick optional

File: packages/react-components/src/components/Drawer/DrawerMenuItem.tsx [hunk 4]
```diff
@@ -66,58 +78,100 @@ export const DrawerMenuItem = (props: _DrawerMenuItemProps): JSX.Element => {
...
 
   const secondaryIcon = props.secondaryIconProps ? (
     <MenuItemIcon {...props.secondaryIconProps} />
   ) : props.subMenuProps ? (
     <MenuItemIcon iconName="ChevronRight" />
   ) : undefined;
 
-  return (
-    <FocusZone shouldFocusOnMount={props.shouldFocusOnMount}>
+  if (props.itemKey === 'reactions') {
```
Comment (JamesBurnside) [inline packages/react-components/src/components/Drawer/DrawerMenuItem.tsx:97]: Instead of bolting this into the existing DrawerMenuItem and coupling a drawer menu with a specific use case, could this be its own component `ReactionsDrawerMenuItem`

File: packages/react-components/src/components/Drawer/DrawerMenu.tsx (hunk 3 above)
Comment (JamesBurnside) [reply packages/react-components/src/components/Drawer/DrawerMenu.tsx:141]: > I think the only difference is not passing in an onItemClick? in which case just make the itemClick optional

Although based on a difference comment I just made about separting the DrawerMenuItem from reactions, the change become something like:

```diff
-         {menuItemsToRender?.slice(0, 1).map((item) => (
-            <DrawerMenuItem
-              {...item}
-              key={`${item.itemKey}` + '0'}
-              shouldFocusOnMount={false}
-              styles={modifiedFirstItemStyle}
-            />
-          ))}
+        {menuItemsToRender?.slice(0, 1).map((item) => (
+            itemKey === 'reactions' ?
+               <ReactionDrawerMenuItem {props} /> :
+               <DrawerMenuItem {props} />;
```

File: packages/react-components/src/components/Drawer/DrawerMenu.tsx (hunk 2 above)
Comment (JamesBurnside) [reply packages/react-components/src/components/Drawer/DrawerMenu.tsx:107]: Then the parent component would use like this:
```tsx
<DrawerMenu
    items={[{
      itemKey: 'myKey1',
      onRender: <ReactionDrawerMenuItem />;
    }, {
      //some other item
      itemKey: 'myKey2',
      onClick: () => {}
    }, //etc
   ]}
/>
```

File: packages/react-components/src/components/styles/ReactionButton.styles.ts [hunk 5]
```diff
@@ -82,3 +82,49 @@ export const reactionToolTipHostStyle = (): ITooltipHostStyles => {
...
+ * @private
+ */
+export const mobileViewMenuItemStyle = (): React.CSSProperties => {
+  return {
+    display: 'flex',
+    justifyContent: 'space-between',
+    alignItems: 'center',
+    flexDirection: 'row',
+    width: '100%',
+    height: '42px'
```
Comment (JamesBurnside) [inline packages/react-components/src/components/styles/ReactionButton.styles.ts:98]: use rem instead of px

File: packages/react-components/src/components/styles/ReactionButton.styles.ts [hunk 6]
```diff
@@ -82,3 +82,49 @@ export const reactionToolTipHostStyle = (): ITooltipHostStyles => {
...
+ *
+ * @private
+ */
+export const mobileViewEmojiStyles = (backgroundImage: string, animationPlayState: string): React.CSSProperties => {
+  const imageResourceUrl = `url(${backgroundImage})`;
+  return {
+    display: 'flex',
+    flexDirection: 'column',
+    height: '100%',
+    width: '44px',
```
Comment (JamesBurnside) [inline packages/react-components/src/components/styles/ReactionButton.styles.ts:115]: use rem instead of px

File: packages/react-components/src/components/styles/ReactionButton.styles.ts [hunk 7]
```diff
@@ -82,3 +82,49 @@ export const reactionToolTipHostStyle = (): ITooltipHostStyles => {
...
+
+/* @conditional-compile-remove(reaction) */
+/**
+ *
+ * @private
+ */
+export const mobileViewMenuItemStyle = (): React.CSSProperties => {
+  return {
+    display: 'flex',
+    justifyContent: 'space-between',
```
Comment (JamesBurnside) [inline packages/react-components/src/components/styles/ReactionButton.styles.ts:94]: on landscape - does this mean they are all really far apart?

**Review by JamesBurnside — CHANGES_REQUESTED** [review]
Blocking item is to remove `"@azure/communication-calling"` from ui-component packlet + other suggestions that are more non-blocking

File: packages/react-components/package.json (hunk 1 above)
Comment (Mohtasim) [reply packages/react-components/package.json:65]: We already have a `Reaction` type in the `react-component`. Having another would create conflict. So importing from the original calling lib. Can I have some pointer how to resolve this?

**Comment by dmceachernmsft** [comment]
image
looks like there might be a need for more padding. the surprised reaction looks a little close to the edge?

File: packages/react-components/package.json (hunk 1 above)
Comment (Mohtasim) [reply packages/react-components/package.json:65]: the expected signature for both types are different.

Comment (JamesBurnside) [reply packages/react-components/package.json:65]: The parent using the UI component then has to map whatever the component type is to its function, e.g.
```tsx
const myFunction = (varStr: 'string') => { console.log(varStr) };

interface ComponentProps: {
  doSomething: (varNum: number) => void;
}

<Component
  // do something requires number, but myFunction want a string, hence we need to convert the
  // arg from doSomething to match the expected input to myFunction, in this case a numToStr function
  doSomething={(varNum) => myFunction(numToStr(varNum))}
/>

```

File: packages/react-components/src/components/styles/ReactionButton.styles.ts (hunk 7 above)
Comment (Mohtasim) [reply packages/react-components/src/components/styles/ReactionButton.styles.ts:94]: yes. not really really far apart, but thats the design, I guess.

File: packages/react-components/package.json (hunk 1 above)
Comment (Mohtasim) [reply packages/react-components/package.json:65]: Thanks. got it. Pushed a new commit addressing the concern.

File: packages/react-components/src/components/Drawer/DrawerMenu.tsx (hunk 2 above)
Comment (Mohtasim) [reply packages/react-components/src/components/Drawer/DrawerMenu.tsx:107]: Having onRenderMenuItem would again force us to implement similar logic on drawerMenuItem. 
Even if we have a separate reaction drawer menu component, we would still need to decide if the item renders reaction menu item or not.

File: packages/react-components/src/components/Drawer/DrawerMenuItem.tsx [hunk 8]
```diff
@@ -66,58 +76,100 @@ export const DrawerMenuItem = (props: _DrawerMenuItemProps): JSX.Element => {
...
-              }}
+        <div style={mobileViewMenuItemStyle()}>
+          {emojis.map((emoji, index) => {
+            const resourceUrl = emojiResource.get(emoji.toString());
+
+            return (
+              <IconButton
+                key={index}
+                onClick={() => {
+                  if (props.onReactionClick !== undefined) {
```
Comment (JamesBurnside) [inline packages/react-components/src/components/Drawer/DrawerMenuItem.tsx:115]: nit: handy javascript thing, you can remove this `if` by using `?`
```
props.onReactionClick?.()
```

File: packages/react-components/src/components/Drawer/DrawerMenu.tsx [hunk 9]
```diff
@@ -104,28 +104,34 @@ export const _DrawerMenu = (props: _DrawerMenuProps): JSX.Element => {
...
-          <DrawerMenuItem
-            {...item}
-            key={`${item.itemKey}` + `${i + 1}`}
-            onItemClick={(ev, itemKey) => {
-              onItemClick(item, ev, itemKey);
-            }}
-          />
-        ))}
-      </Stack>
+      {
```
Comment (JamesBurnside) [inline packages/react-components/src/components/Drawer/DrawerMenu.tsx:107]: can remove this again now?

File: packages/react-components/src/components/Drawer/DrawerMenuItem.tsx [hunk 10]
```diff
@@ -66,58 +76,103 @@ export const DrawerMenuItem = (props: _DrawerMenuItemProps): JSX.Element => {
...
+          {emojis.map((emoji, index) => {
+            const resourceUrl = emojiResource.get(emoji.toString());
+
+            return (
+              <IconButton
+                key={index}
+                onClick={() => {
+                  props.onReactionClick?.(emoji);
+                }}
+                style={mobileViewEmojiStyles(resourceUrl ? resourceUrl : '', 'running')}
```
Comment (JamesBurnside) [inline packages/react-components/src/components/Drawer/DrawerMenuItem.tsx:121]: js tip: can use null coalescence operator here:
```suggestion
                style={mobileViewEmojiStyles(resourceUrl ?? '', 'running')}
```

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx [hunk 11]
```diff
@@ -319,26 +337,30 @@ export const MoreDrawer = (props: MoreDrawerProps): JSX.Element => {
         itemKey: 'dynamicSelectionKey',
         text: localeStrings.strings.call.moreButtonGalleryFloatingLocalLayoutLabel,
         onItemClick: () => {
+          /* @conditional-compile-remove(gallery-layout-composite) */
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:340]: this looks unrelated to reactions - was this needed for a build fix?

File: change-beta/@azure-communication-react-620cfff7-2075-4461-b442-1363c891c420.json [hunk 12]
```diff
@@ -0,0 +1,9 @@
+{
+  "type": "prerelease",
+  "area": "improvement",
+  "workstream": "removing unwanted dependency",
```
Comment (JamesBurnside) [inline change-beta/@azure-communication-react-620cfff7-2075-4461-b442-1363c891c420.json:4]: `workstream` is the feature workstream, in this case it would be `reactions`. Its just used for grouping like changes in the changelog generation

File: change-beta/@azure-communication-react-d54219e1-a29e-44dc-a5ba-a9a4e4193856.json [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (JamesBurnside) [inline change-beta/@azure-communication-react-d54219e1-a29e-44dc-a5ba-a9a4e4193856.json:1]: one change file per PR is sufficient (unless there is a time when multiple things are introduced in the same PR, but then it really should be multiple PRs)

File: packages/react-components/src/components/Drawer/DrawerMenuItem.tsx (hunk 4 above)
Comment (Mohtasim) [reply packages/react-components/src/components/Drawer/DrawerMenuItem.tsx]: Thanks for the suggestion and guidance. :)

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx (hunk 11 above)
Comment (Mohtasim) [reply packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:347]: Yes. Otherwise it wasn't passing the CI job checks.

File: change-beta/@azure-communication-react-620cfff7-2075-4461-b442-1363c891c420.json (hunk 12 above)
Comment (Mohtasim) [reply change-beta/@azure-communication-react-620cfff7-2075-4461-b442-1363c891c420.json:4]: I have been using work stream in the wrong way this whole time ⭕

File: packages/react-components/src/components/Drawer/DrawerMenuItem.tsx [hunk 14]
```diff
@@ -54,6 +56,12 @@ export interface _DrawerMenuItemProps {
    * Property to set the focus since this is the first item in the menu
    */
   shouldFocusOnMount?: boolean;
+  /* @conditional-compile-remove(reaction) */
+  reactionResources?: ReactionResources;
```
Comment (JamesBurnside) [inline packages/react-components/src/components/Drawer/DrawerMenuItem.tsx:60]: can remove these now

Comment (Mohtasim) [reply packages/react-components/src/components/Drawer/DrawerMenuItem.tsx:60]: No. Because ReactionDrawerItem is using the same prop. It does not have a separate props.

Comment (JamesBurnside) [reply packages/react-components/src/components/Drawer/DrawerMenuItem.tsx:60]: We pass the reactions in as a custom component, the DrawerMenuItem doesn't use anything reaction so shouldn't be on its interface. We can create a `interface _ReactionDrawerMenuItemProps extends _DrawerMenuItemProps { reactionsRelatedProps }`

Comment (JamesBurnside) [reply packages/react-components/src/components/Drawer/DrawerMenuItem.tsx:60]: (even then _ReactionDrawerMenuItemProps doesn't need to extend the _DrawerMenuItemProps, it can just have its own interface for whatever props it needs)

Comment (Mohtasim) [reply packages/react-components/src/components/Drawer/DrawerMenuItem.tsx:60]: Yes. We can remove them. Thanks :)

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx [hunk 15]
```diff
@@ -319,26 +340,30 @@ export const MoreDrawer = (props: MoreDrawerProps): JSX.Element => {
...
           styles: { root: { lineHeight: 0 } }
         },
+        /* @conditional-compile-remove(gallery-layout-composite) */
         secondaryIconProps: props.userSetGalleryLayout === 'floatingLocalVideo' ? { iconName: 'Accept' } : undefined
       },
       {
         itemKey: 'focusedContentSelectionKey',
         text: localeStrings.strings.call.moreButtonGalleryFocusedContentLayoutLabel,
         onItemClick: () => {
+          /* @conditional-compile-remove(gallery-layout-composite) */
```
Comment (dmceachernmsft) [inline packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:358]: you are removing released features here. this should not be added in any of the places you added it

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx [hunk 16]
```diff
@@ -319,26 +340,30 @@ export const MoreDrawer = (props: MoreDrawerProps): JSX.Element => {
...
           props.onUserSetGalleryLayout && props.onUserSetGalleryLayout('floatingLocalVideo');
           onLightDismiss();
         },
         iconProps: {
           iconName: 'FloatingLocalVideoGalleryLayout',
           styles: { root: { lineHeight: 0 } }
         },
+        /* @conditional-compile-remove(gallery-layout-composite) */
```
Comment (dmceachernmsft) [inline packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:351]: This code has been GA'd

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx [hunk 17]
```diff
@@ -349,6 +374,7 @@ export const MoreDrawer = (props: MoreDrawerProps): JSX.Element => {
     itemKey: 'defaultSelectionKey',
     text: localeStrings.strings.call.moreButtonGalleryDefaultLayoutLabel,
     onItemClick: () => {
+      /* @conditional-compile-remove(gallery-layout-composite) */
```
Comment (dmceachernmsft) [inline packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:377]: this needs to stay for the composite

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx (hunk 11 above)
Comment (dmceachernmsft) [reply packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:343]: This is a GA'd feature part of gallery layouts

**Review by dmceachernmsft — CHANGES_REQUESTED** [review]
We need to make sure that the icon you are removing is there. please remove the CC that you added to the gallery section of the drawer. If you need help working out the error please reach out to me.

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx (hunk 11 above)
Comment (Mohtasim) [reply packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:343]: But there are other parts where we have cc ... those were not removed. so it ends up bubbling up errors. For the sake of successful I am adding this. Removing them should be a different KTLO item.

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx (hunk 17 above)
Comment (Mohtasim) [reply packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:377]: But there are other parts where we have cc ... those were not removed. so it ends up bubbling up errors. For the sake of successful I am adding this. Removing them should be a different KTLO item.

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx (hunk 16 above)
Comment (Mohtasim) [reply packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:351]: But there are other parts where we have cc ... those were not removed. so it ends up bubbling up errors. For the sake of successful I am adding this. Removing them should be a different KTLO item.

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx (hunk 15 above)
Comment (Mohtasim) [reply packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:358]: But there are other parts where we have cc ... those were not removed. so it ends up bubbling up errors. For the sake of successful I am adding this. Removing them should be a different KTLO item.

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx (hunk 17 above)
Comment (Mohtasim) [reply packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx:377]: In the MoreDrawer File there are 10 other occurences which was there and I did not add or remove them. Without adding these 4, I get build error for stable. Lets make a separate task to remove those CC.

**Review by dmceachernmsft — APPROVED** [review]

File: packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx (hunk 11 above)
Comment (Mohtasim) [reply packages/react-composites/src/composites/common/Drawer/MoreDrawer.tsx]: Thanks @dmceachernmsft , looks like removing line 21 and others solved the issue. Removing them from all occurances.
